doc/monads/monads.lean

```diff
@@ -0,0 +1,307 @@
+/-!
+# Monads
+
+Building on [Functors](functors.lean.md) and [Applicatives](applicatives.lean.md) we can now introduce
+[monads](https://en.wikipedia.org/wiki/Monad_%28category_theory%29).
+
+A monad is another type of abstract, functional structure. Let's explore what makes it different
+from the first two structures.
+
+## What is a Monad?
+
+A monad is a computational context. It provides a structure that allows you to chain together
+operations that have some kind of shared state or similar effect. Whereas pure functional code can
+only operate on explicit input parameters and affect the program through explicit return values,
+operations in a monad can affect other computations in the chain implicitly through side effects,
+especially modification of an implicitly shared value.
+
+## How are monads represented in Lean?
+
+Like functors and applicatives, monads are represented with a type class in Lean:
+
+```lean,ignore
+class Monad (m : Type u → Type v) extends Applicative m, Bind m where
+```
+
+Just as every applicative is a functor, every monad is also an applicative and
+there's one more new base type class used here that you need to understand, namely, `Bind`.
+
+```lean,ignore
+class Bind (f : Type u → Type v) where
+  bind : {α β : Type u} → f α → (α → f β) → f β
+```
+
+The `bind` operator also has infix notation `>>=` where `x >>= g` represents the result of
+executing `x` to get a value of type `f α` then unwrapping the value `α` from that and
+passing it to function `g` of type `α → f β` returning the result of type `f β` where
+`f` is the target structure type (like `Option` or List)
+
+This `bind` operation looks similar to the other ones you've seen so far, if you put
+them all together `Monad` has the following operations:
+
+```lean,ignore
+class Monad (f : Type u → Type v) extends Applicative f, Bind f where
+  pure {α : Type u} : α → f α
+  map : {α β : Type u} → (α → β) → f α → f β
+  seq : {α β : Type u} → f (α → β) → (Unit → f α) → f β
+  bind : {α β : Type u} → f α → (α → f β) → f β
+  ...
+```
+
+Notice `Monad` also contains `pure` it must also have a "default" way to wrap a value in the
+structure.
+
+The `bind` operator is similar to the applicative `seq` operator in that it chains two operations,
+with one of them being function related.
+
+Notice that `bind`, `seq` and `map` all take a function of some kind, and then a structure wrapped
```

Comment:
This paragraph looks weird to me. Not sure what is the point you are trying to make here. Does it come from the source material?

Reply:
Yes this is the original, I only added type information so you can map the prose back to the Monad type class definition shown a few lines above:

> All these operators take a function of some kind, and then a structure wrapped over an a type, and then produce a structure wrapping a b type. They just vary in what the function looks like. For the functor, the function is a normal pure function. For applicatives, the function is still pure, but wrapped in the structure. Now with monads, our function argument takes a "pure" input but produces an output in the structure.

Reply:
I personally found it helpful to look at the 3 different functions for map, seq, bind and understand their differences, which this paragraph does, and it helped me get familiar with the language "structure wrapped over a type" that is being performed by the `f α` and  `f β` terms in the Monad type class definition.

Reply:
The paragraph seems to be stating in English the information that is in their types. I don't see it providing new information. The rest of the text is already showing how to use them, the laws they obey, and examples.

Reply:
ha ha but Leo that's because you think and breath Lean.  The rest of us need it spelled out so we can understand what all those symbols mean!

Reply:
In my point of view, you are just spelling out the types of these functions. Are you assuming that people reading this section cannot read Lean types?

Reply:
Are you saying you don't like the original paragraph either, or are you saying you don't like the type annotations I added?

Reply:
I checked the original. Note that they flipped the `bind` to make their point. I also think their sentence is a bit weird, but it reads better.

Reply:
I couldn't find =<< in Lean?  So what's your conclusion?  Delete the whole paragraph, leave it, remove the types, rewrite it?  I'm happy with either decision.

Reply:
How about this instead:


Notice that `bind`, `seq` and `map` all take a function of some kind.  Let's examine those function types:

- map: `(α → β)`
- seq: `f (α → β)`   -- ignoring `(Unit → f α)` for now.
- bind: `(α → f β)`

So `map` is a pure function, `seq` is a pure function wrapped in the structure, and `bind` takes
a "pure" input but produces an output wrapped in the structure.

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/11ceaf23f12fda8daf3ae64f991011672ad8409c (sorry about the reflow, I'm using a vscode reflow line wrapping plugin and I didn't expect it to roll forwards like that).

Reply:
> In my point of view, you are just spelling out the types of these functions. Are you assuming that people reading this section cannot read Lean types?

I agree with @lovettchris 's point on this. The purpose of documentation is to help people understand, so restating information that is "already contained in the types" is not redundant information if it helps people understand how to gloss lean types in general. In particular, `bind` has a famously complex type signature that has birthed many tutorials and articles, even though it's quite trivial once you are used to this kind of thing.